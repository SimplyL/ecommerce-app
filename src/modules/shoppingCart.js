const ADD = 'ecommerce-app/shoppingCart/ADD';
const REMOVE = 'ecommerce-app/shoppingCart/REMOVE';
const CLEAR = 'ecommerce-app/shoppingCart/CLEAR';

const reducer = (state = [], action = {}) => {
  switch (action.type) {
    case ADD: {
      const { id, quantity } = action.payload;
      const existingProduct = state.find(item => item.id === id);
      let items = [];

      if (existingProduct) {
        existingProduct.quantity = existingProduct.quantity + quantity;
        items = [
          ...state.filter(item => item.id !== id),
          existingProduct,
        ];
      } else {
        items = [
          ...state,
          {
            id,
            quantity,
          }
        ];
      }

      return items;
    }

    case REMOVE: {
      return state.filter(item => item.id !== action.payload);
    }

    case CLEAR: {
      return [];
    }

    default: return state;
  }
}

export const addProduct = payload => dispatch => (
  dispatch({ type: ADD, payload })
);

export const removeProduct = payload => dispatch => (
  dispatch({ type: REMOVE, payload })
);

export const clearCart = payload => dispatch => (
  dispatch({ type: CLEAR, payload })
);

export default reducer;
