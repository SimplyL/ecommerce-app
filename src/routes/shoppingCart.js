import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import ProductItem from '../components/productItem';
import { productType } from '../constants/types';

const renderCartItem = products => item => {
  const cartItem = products.find(product => product.id === item.id);

  return (
    <ProductItem
      key={cartItem.id}
      product={cartItem}
      quantity={item.quantity}
    />
  );
};

class ShoppingCart extends Component {
  render() {
    const { shoppingCart, products } = this.props;
    return (
      <div>
        <div>Shopping cart page</div>
        <div>
          {shoppingCart.map(renderCartItem(products))}
        </div>
      </div>

    );
  }
};

ShoppingCart.propTypes = {
  products: PropTypes.arrayOf(productType).isRequired,
}

const mapStateToProps = state => ({ shoppingCart: state.shoppingCart });

export default connect(mapStateToProps)(ShoppingCart);
