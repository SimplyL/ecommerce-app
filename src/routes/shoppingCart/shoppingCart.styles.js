import styled from 'styled-components';
import { device } from '../../App.styles';

const Container = styled.div`
  display: grid;
`;
Container.displayName = 'Container';

const CheckoutButton = styled.button`
  padding: 10px 20px 10px 20px;
  background-color: #ffffff;
  width: 150px;
  color: #3080e8;
  border: 1px solid #3080e8;
  border-radius: 4px;
  text-align: center;
  text-decoration: none;
  font-size: 1.1em;
  box-sizing: border-box;
  margin-top: 20px

  @media ${device.tablet} {
    margin-top: 0px;
    justify-self: end;
  }

  &:hover {
    color: #ffffff;
    background-color: #3080e8;
    cursor: pointer;
  }
`;
CheckoutButton.displayName = 'CheckoutButton';

const CheckoutContainer = styled.div`
  border: 1px solid #F0F0F0;
  border-radius: 5px;
  padding: 20px;
  display: grid;
  justify-items: center;
  align-items: center;

  @media ${device.tablet} {
    grid-template-columns: 1fr auto;
  }
`;
CheckoutContainer.displayName = 'CheckoutContainer';

const Label = styled.span`
  font-weight: 600;
`;
Label.displayName = 'Label';

export {
  Container,
  CheckoutButton,
  CheckoutContainer,
  Label,
}