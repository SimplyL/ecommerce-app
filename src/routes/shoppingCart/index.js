import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import CartItem from '../../components/productItem/cartItem';
import { PairContainer, PairValue } from '../../components/productItem/productItem.styles';
import Title from '../../components/title';
import { PRODUCT_TYPE } from '../../constants/types';
import { clearCart } from '../../modules/shoppingCart';
import {
  Container,
  CheckoutButton,
  CheckoutContainer,
  Label,
} from './shoppingCart.styles';

const renderCartItem = products => item => {
  const cartItem = products.find(product => product.id === item.id);

  return (
    <CartItem
      key={cartItem.id}
      product={cartItem}
      quantity={item.quantity}
    />
  );
};

const calculateFullPrice = (shoppingCart, products) => {
  const fullPrice = shoppingCart.map(item =>
    products.find(product => product.id === item.id).price * item.quantity
  ).reduce((acc, curr) => acc + curr);

  return parseFloat(fullPrice).toFixed(2);
}

export const ShoppingCart = ({ shoppingCart, products, clearCart }) => (
  <>
    <Title name="Shopping cart" />
    {
      !!shoppingCart.length ? (
        <Container>
          {shoppingCart.map(renderCartItem(products))}
          <CheckoutContainer>
            <PairContainer>
              <Label>Full price:</Label>
              <PairValue>
                {calculateFullPrice(shoppingCart, products)} &euro;
              </PairValue>
            </PairContainer>
            <CheckoutButton onClick={clearCart}>Checkout</CheckoutButton>
          </CheckoutContainer>
        </Container>
      ) : (<div>Your cart is empty</div>)
    }
  </>

);

ShoppingCart.propTypes = {
  products: PropTypes.arrayOf(PRODUCT_TYPE).isRequired,
  clearCart: PropTypes.func.isRequired,
  shoppingCart: PropTypes.arrayOf(
    PropTypes.object.isRequired,
  ),
}

const mapStateToProps = state => ({ shoppingCart: state.shoppingCart });

export default connect(mapStateToProps, { clearCart })(ShoppingCart);
