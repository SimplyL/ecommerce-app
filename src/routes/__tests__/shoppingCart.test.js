import React from 'react';
import { shallow } from 'enzyme';
import { ShoppingCart } from '../shoppingCart';
import mockProducts from './__mocks__/products';

describe('Shopping cart page', () => {
  let wrapper;
  const shoppingCartMock = [
    {
      id: '0',
      quantity: 4,
    },
    {
      id: '1',
      quantity: 1,
    }
  ];
  const mockFn = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<ShoppingCart
      products={mockProducts}
      shoppingCart={shoppingCartMock}
      clearCart={mockFn}
    />);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render cart list with correct item number', () => {
    // ConnectFunction hack due to lack of Enzyme support for current redux implementation.
    expect(wrapper.find('ConnectFunction').length).toEqual(2);
  });

  it('should call checkout function', () => {
    wrapper.find('CheckoutButton').simulate('click');
    expect(mockFn).toHaveBeenCalled();
  });
});
