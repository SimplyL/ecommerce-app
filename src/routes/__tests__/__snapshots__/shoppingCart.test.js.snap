// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping cart page should render without crashing 1`] = `
<Fragment>
  <Title
    name="Shopping cart"
  />
  <Container>
    <ConnectFunction
      key="0"
      product={
        Object {
          "description": "Test description",
          "id": "0",
          "image": "roxy-sugar-banana.jpg",
          "name": "Model",
          "price": 4.99,
        }
      }
      quantity={4}
    />
    <ConnectFunction
      key="1"
      product={
        Object {
          "description": "Test",
          "id": "1",
          "image": "gnu-fb-head.jpg",
          "name": "Model",
          "price": 49.99,
        }
      }
      quantity={1}
    />
    <CheckoutContainer>
      <PairContainer>
        <Label>
          Full price:
        </Label>
        <styled.span>
          69.95
           â‚¬
        </styled.span>
      </PairContainer>
      <CheckoutButton
        onClick={[MockFunction]}
      >
        Checkout
      </CheckoutButton>
    </CheckoutContainer>
  </Container>
</Fragment>
`;
